File: java/src/Analysis.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.Set;

public class Analysis {
  private ArrayList<Integer> analysisList;

  Analysis() {
    analysisList = new ArrayList<Integer>();
  }

  public void Load(int[] data) {
    for (int i = 0; i < data.length; i++) {
      analysisList.add(data[i]);
    }
  }

  public int Sum() {
    if (analysisList.isEmpty()) {
      throw new Error("No data to analyze");
    } else {
      int sum = 0;
      for (int i = 0; i < analysisList.size(); i++) {
        sum += analysisList.get(i);
      }
      return sum;
    }
  }

  public int Sum(int key) {
    if (analysisList.isEmpty()) {
      throw new Error("No data to analyze");
    } else {
      int sum = 0;
      for (int i = 0; i < analysisList.size(); i++) {
        if (analysisList.get(i) == key) {
          sum += analysisList.get(i);
        }
      }
      return sum;
    }
  }

  public int Count() {
    if (analysisList.isEmpty()) {
      throw new Error("No data to analyze");
    } else {
      return analysisList.size();
    }
  }

  public int Count(int key) {
    if (analysisList.isEmpty()) {
      throw new Error("No data to analyze");
    } else {
      int count = 0;
      for (int i = 0; i < analysisList.size(); i++) {
        if (analysisList.get(i) == key) {
          count++;
        }
      }
      return count;
    }
  }

  public int Min() {
    if (analysisList.isEmpty()) {
      throw new Error("No data to analyze");
    } else {
      int min = analysisList.get(0);
      for (int i = 0; i < analysisList.size(); i++) {
        if (analysisList.get(i) < min) {
          min = analysisList.get(i);
        }
      }
      return min;
    }
  }

  public int Max() {
    if (analysisList.isEmpty()) {
      throw new Error("No data to analyze");
    } else {
      int max = analysisList.get(0);
      for (int i = 0; i < analysisList.size(); i++) {
        if (analysisList.get(i) > max) {
          max = analysisList.get(i);
        }
      }
      return max;
    }
  }

  public double Average() {
    if (analysisList.isEmpty()) {
      throw new Error("No data to analyze");
    } else {
      return (double) Sum() / analysisList.size();
    }
  }

  public int Range() {
    if (analysisList.isEmpty()) {
      throw new Error("No data to analyze");
    } else {
      return Max() - Min();
    }
  }

  public double Mean() {
    if (analysisList.isEmpty()) {
      throw new Error("No data to analyze");
    } else {
      return Math.round((double) Sum() / (double) Count() * 100.0) / 100.0;
    }
  }

  public double Median() {
    if (analysisList.isEmpty()) {
      throw new Error("No data to analyze");
    } else {
      ArrayList<Integer> data = analysisList;
      data.sort(Comparator.naturalOrder());
      int dataLength = data.size();
      if (dataLength % 2 == 0) {
        return (double) (data.get(dataLength / 2) + data.get(dataLength / 2 - 1)) / 2.0;
      } else {
        return data.get(dataLength / 2);
      }
    }
  }

  public String Mode() {
    if (analysisList.isEmpty()) {
      throw new Error("No data to analyze");
    } else {
      Set<Integer> keys = Set.copyOf(analysisList);
      int mode = 0, count = 0;
      for (int key : keys) {
        int keyCount = Count(key);
        if (keyCount > count) {
          mode = key;
          count = keyCount;
        }
      }
      return ("{ mode: " + mode + ", count: " + count + " }");
    }
  }

  public double Variance() {
    if (analysisList.isEmpty()) {
      throw new Error("No data to analyze");
    } else {
      double variance = 0;
      for (Integer number : analysisList) {
        variance += Math.pow(number - Mean(), 2);
      }
      return Math.round(variance / Count() * 10000.0) / 10000.0;
    }
  }

  public double StandardDeviation() {
    if (analysisList.isEmpty()) {
      throw new Error("No data to analyze");
    } else {
      return Math.round(Math.sqrt(Variance()) * 10000.0) / 10000.0;
    }
  }

  public int[][] FrequencyDistribution() {
    if (analysisList.isEmpty()) {
      throw new Error("No data to analyze");
    } else {
      Set<Integer> keys = Set.copyOf(analysisList);
      int[][] frequencyDistribution = new int[keys.size()][2];
      int index = 0;
      for (int key : keys) {
        frequencyDistribution[index][0] = key;
        frequencyDistribution[index][1] = Count(key) * 5;
        index++;
      }
      Arrays.sort(frequencyDistribution, (a, b) -> Double.compare(b[1], a[1]));
      return frequencyDistribution;
    }
  }

  public void Describe() {
    if (analysisList.isEmpty()) {
      throw new Error("No data to analyze");
    } else {
      System.out.println("Sum: " + Sum());
      System.out.println("Count: " + Count());
      System.out.println("Min: " + Min());
      System.out.println("Max: " + Max());
      System.out.println("Range: " + Range());
      System.out.println("Mean: " + Mean());
      System.out.println("Median: " + Median());
      System.out.println("Mode: " + Mode());
      System.out.println("Variance: " + Variance());
      System.out.println("Standard Deviation: " + StandardDeviation());
      System.out.println("Frequency Distribution: " + Arrays.deepToString(FrequencyDistribution()));
    }
  }
}

File: python/Analysis.py
class Analysis:
  def __init__(self):
    self._data = list()
  
  def count(self,key=None):
    if not self._data:  
      raise ValueError("Error: No data loaded")
    if key:
      return self._data.count(key)
    return len(self._data)
  
  def sum(self, key=None):
    if not self._data:  
      raise ValueError("Error: No data loaded")
    if not key:
      total = 0
      for i in range(len(self._data)):
        total+=self._data[i]
      return total
    if key:
      total = 0
      for i in range(len(self._data)):
        if key == self._data[i]:
          total+=self._data[i]
      return total

  def min(self):
    if not self._data:  
      raise ValueError("Error: No data loaded")
    min = self._data[0]
    for i in range(len(self._data)):
      if min > self._data[i]:
        min = self._data[i]
    return min
  
  def max(self):
    if not self._data:  
      raise ValueError("Error: No data loaded")
    max = self._data[0]
    for i in range(len(self._data)):
      if max < self._data[i]:
        max = self._data[i]
    return max
  
  def range(self):
    if not self._data:  
      raise ValueError("Error: No data loaded")
    return self.max() - self.min()
  
  def mean(self):
    if not self._data:  
      raise ValueError("Error: No data loaded")
    return round(self.sum() / self.count(),2)
  
  def get_sorted_data(self):
    if not self._data:  
      raise ValueError("Error: No data loaded")
    return sorted(self._data)

  def median(self):
    if not self._data:  
      raise ValueError("Error: No data loaded")
    data = self.get_sorted_data()
    data_len = len(data)
    if data_len % 2 == 0:
      return (data[data_len // 2 - 1] + data[data_len // 2]) / 2
    return data[data_len // 2]
  
  def mode(self):
    if not self._data:  
      raise ValueError("Error: No data loaded")
    mode = { "mode": 0, "count": 0 }
    keys = set(self._data)
    for key in keys:
      count = self.count(key)
      if(mode["count"] < count):
        mode["mode"] = key
        mode["count"] = count
    return mode
  
  def variance(self):
    if not self._data:  
      raise ValueError("Error: No data loaded")
    variance = 0
    for i in range(len(self._data)):
      variance += pow(self._data[i] - self.mean(),2)
    return round(variance / self.count(),2)
  
  def standard_deviation(self):
    if not self._data:  
      raise ValueError("Error: No data loaded")
    return round(pow(self.variance(),.5),2)
  
  def frequency_distribution(self):
    if not self._data:  
      raise ValueError("Error: No data loaded")
    freqDist = list()
    keys = set(self._data)
    for key in keys:
      count  = self.count(key)
      freqDist.append((key,count*5))
    freqDist.sort(key=lambda a: a[1],reverse=True)
    return freqDist

  def describe(self):
    if not self._data:  
      raise ValueError("Error: No data loaded")
    print("Count:", self.count());
    print("Sum: ", self.sum());
    print("Min: ", self.min());
    print("Max: ", self.max());
    print("Range: ", self.range());
    print("Mean: 3".format(self.mean()));
    print("Median: ", self.median());
    print("Mode: ", self.mode());
    print("Standard Deviation: ".format(self.standard_deviation()));
    print("Variance: ".format(self.variance()));
    print("Frequency Distribution: ", self.frequency_distribution())

  def load(self, data):
    self._data = data.copy()
